### PR TITLE
CASSANDRA-19635 - Run integration tests with C* 5.x

### PR DESCRIPTION
**Pull request by lukasz-antoniak**
JIRA link: https://issues.apache.org/jira/browse/CASSANDRA-19635

PR contains necessary changes to run integration tests with Apache Cassandra 5.0-beta1. Developers can run the tests using:
```
$ mvn clean install -Dccm.version=5.0-beta1 -DtestJavaHome=/opt/java17/home -Ptest-jdk-17
```
Make sure you have latest CCM version installed locally and available in `$PATH`.

### CODE REVIEW

File: test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/CcmBridge.java [hunk 1]
```diff
@@ -199,7 +199,7 @@ public Version getCassandraVersion() {
   private String getCcmVersionString(Version version) {
     // for 4.0 pre-releases, the CCM version string needs to be "4.0-alpha1" or "4.0-alpha2"
     // Version.toString() always adds a patch value, even if it's not specified when parsing.
-    if (version.getMajor() == 4
```
Comment (tolbertam) [inline test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/CcmBridge.java:202]: may be more future-proof to just say `>= 4` here as likely to have same problem with future releases.

File: test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/SchemaChangeSynchronizer.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+package com.datastax.oss.driver.api.testinfra.ccm;
+
+import java.util.concurrent.Semaphore;
+
+/**
+ * Running multiple parallel integration tests may fail due to query timeout when trying to apply
+ * several schema changes at once. Limit concurrently executed DDLs to 5.
+ */
+public class SchemaChangeSynchronizer {
+  private static final Semaphore lock = new Semaphore(5);
```
Comment (tolbertam) [inline test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/SchemaChangeSynchronizer.java:27]: Any reason why we shouldn't just set this to 1?

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java [hunk 3]
```diff
@@ -34,23 +34,27 @@
...
 import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.ClassRule;
 import org.junit.Test;
-import org.junit.experimental.categories.Category;
 import org.junit.rules.RuleChain;
 import org.junit.rules.TestRule;
 
-@Category(ParallelizableTests.class)
+// Do not run LWT tests in parallel because they may interfere. Tests operate on the same row.
```
Comment (tolbertam) [inline integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java:50]: comment isn't applicable right?  Probably can remove it as I think the annotation description describes it well.

**Review by tolbertam — COMMENTED** [review]
Just a few minor comments/suggestions.  This seems to be working good for me when I use JDK 17 as my primary JVM.

I'm still running through the suite locally, but leaving feedback early as I wait on that to complete.

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java (hunk 3 above)
Comment (tolbertam) [reply integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java:50]: ah I see, there are LWT specific tests in here, so this comment is actually applicable (`should_delete_if_exists_reactive` for example)

File: test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/BaseCcmRule.java [hunk 4]
```diff
@@ -38,7 +38,7 @@ public abstract class BaseCcmRule extends CassandraResourceRule {
             new Thread(
                 () -> {
                   try {
-                    ccmBridge.remove();
+                    ccmBridge.close();
```
Comment (tolbertam) [inline test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/BaseCcmRule.java:41]: Why the change from `remove` to `close`?  Possible this was changed to check directory outputs after but forgot to change back?

File: test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/SchemaChangeSynchronizer.java (hunk 2 above)
Comment (lukasz-antoniak) [reply test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/SchemaChangeSynchronizer.java:27]: Still wanted to preserve some parallelism in builds with lower C* versions. Also did not want to make this logic C* version dependent, just to simplify things. If you wish I am happy to change it to 1.

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java (hunk 3 above)
Comment (lukasz-antoniak) [reply integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java:50]: Yes, this was a flaky test that I have fixed while running complete IT set multiple times a day :).

File: test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/BaseCcmRule.java (hunk 4 above)
Comment (lukasz-antoniak) [reply test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/BaseCcmRule.java:41]: I have made `CcmBridge#close()` dependent on `created` flag. Now we execute `ccm remove` only once if the cluster was previously created and not removed after that. If you run FT and log CCM command executions, you would see that we clean-up cluster twice (execute `ccm remove`):

1. In `BaseCcmRule#after()`.
2. As part of `AutoCloseable` and resource `close()` method.

Of course running `ccm remove` twice does not do any harm. IMO this fix is most about taste, so I can revert this back.

Comment (tolbertam) [reply test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/BaseCcmRule.java:41]: Now that you've explained it, it makes sense to me, +1 for keeping it this way.

File: test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/SchemaChangeSynchronizer.java (hunk 2 above)
Comment (tolbertam) [reply test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/SchemaChangeSynchronizer.java:27]: I think that concurrent schema changes happen to 'work' in older versions is probably more by luck than anything.  I think it likely would not slow down things noticeable if we set it to 1, and its a lot more appropriate (have seen too many incidents caused by concurrent schema changes, its always nice to enforce 1 at a time).  That said, i'm cool with merging and leaving it at 5 as well, since it's better than what it does currently :+1:

**Review by tolbertam — APPROVED** [review]
this is excellent, thank you @lukasz-antoniak !

File: test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/CcmBridge.java (hunk 1 above)
Comment (lukasz-antoniak) [reply test-infra/src/main/java/com/datastax/oss/driver/api/testinfra/ccm/CcmBridge.java:202]: Good catch, modified.

File: integration-tests/src/test/java/com/datastax/oss/driver/core/metadata/SchemaIT.java [hunk 5]
```diff
@@ -254,6 +254,20 @@ public void should_get_virtual_metadata() {
...
+                  + "    task_id timeuuid,\n"
+                  + "    completion_ratio double,\n"
+                  + "    kind text,\n"
+                  + "    progress bigint,\n"
+                  + "    sstables int,\n"
+                  + "    target_directory text,\n"
+                  + "    total bigint,\n"
+                  + "    unit text,\n"
+                  + "    PRIMARY KEY (keyspace_name, table_name, task_id)\n"
+                  + "); */",
```
Comment (absurdfarce) [inline integration-tests/src/test/java/com/datastax/oss/driver/core/metadata/SchemaIT.java]: Nit: the sequence of candidate sstable_tasks schemas goes DSE 6.8+/C* 4.0/C* 4.1.  According to that sequence this block should appear last, after C* 4.1.

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java [hunk 6]
```diff
@@ -34,23 +34,27 @@
...
+@BackendRequirement(
+    type = BackendType.CASSANDRA,
+    description = "SASI indexes broken in DSE. See InventoryITBase#BROKEN_SASI_VERSION.")
```
Comment (absurdfarce) [inline integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java]: I think this is too strong a constraint.  IIRC SASI indexes were borked in a specific DSE release.  [The commit](https://github.com/apache/cassandra-java-driver/commit/c94e39c6ef8b45259a38b7e4a9191f40472e193d) which introduced BROKEN_SASI_VERSION has a lot of useful context here, specifically the definition of [isSasiBroken()](https://github.com/apache/cassandra-java-driver/blob/4.x/integration-tests/src/test/java/com/datastax/oss/driver/mapper/InventoryITBase.java#L95-L99) which checks for an exact DSE version.

**Comment by absurdfarce** [comment]
Agree with @tolbertam ; well done @lukasz-antoniak!  There are a few nits I'd like to see cleaned up but once those are in place I think I'm a 👍 as well.

File: integration-tests/src/test/java/com/datastax/oss/driver/core/metadata/SchemaIT.java (hunk 5 above)
Comment (lukasz-antoniak) [reply integration-tests/src/test/java/com/datastax/oss/driver/core/metadata/SchemaIT.java]: Fixed.

File: integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java (hunk 6 above)
Comment (lukasz-antoniak) [reply integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java]: I had a wrong impression that there are more DSE versions that are actually broken. Hereby integration tests for deletion of entities, we do not want to run in parallel, because they may interfere one with another. Shared `CcmBridge` can be used only with `@Category(ParallelizableTests.class)`, which we do not want. `CustomCcmRule` is mostly used together with `@BackendRequirement`. I had a problem with this test running agains DSE, because `enable_sasi_indexes` is an invalid configuration parameter.

So to make this test work for C*, I had to set `enable_sasi_indexes = true`, but not for DSE.

After further debugging, I have found the root cause - `DefaultCcmBridgeBuilderCustomizer` is not applied to `CustomCcmRule`. I shall not reuse `DefaultCcmBridgeBuilderCustomizer`, because it adds other parameters that causes other integration tests to break (once using `CustomCcmRule`). In the end, I have implemented a dedicated `configureCcm` method inside the tests. `enable_sasi_indexes` is required for one delete statement to succeed inside this test class, so I think we can leave configuration local.

Comment (absurdfarce) [reply integration-tests/src/test/java/com/datastax/oss/driver/mapper/DeleteReactiveIT.java]: This seems fine to me; it fits nicely with the pattern of each test being responsible for whatever customization it needs to operate.

**Review by absurdfarce — APPROVED** [review]
Nice work @lukasz-antoniak!  With your most recent changes I think we're there!

**Comment by absurdfarce** [comment]
Excellent, thanks for making those final changes so quickly @lukasz-antoniak!  With your most recent work I believe we have all the code changes we need to get this merged!

The next step is to consolidate all of these changes down to a single commit.  If it's useful we've provided some [fairly detailed instructions](https://github.com/apache/cassandra-java-driver/pull/1896#issuecomment-2099192505) on how to do this in a comment on a prior PR.  In your case the commit message should read something like the following:

```
CASSANDRA-19635: Run integration tests with C* 5.x
    
patch by Lukasz Antoniak; reviewed by Andy Tolbert, and Bret McGuire for CASSANDRA-19635
```

Feel free to ask questions if any part of that process isn't clear!

**Comment by absurdfarce** [comment]
Yup, this looks good to me.  Thanks @lukasz-antoniak!
